### ISSUE TITLE
[REQUEST] Allow more customization of the status line

### ISSUE THREAD

**Issue by 0phoff** [mawww/kakoune#4445]
### Feature

Coming from VIM, I am used to be able to customize the entire look and feel of my editor, in order to merge it in my terminal/tmux setup nicely. However, Kakoune encodes a lot of very useful information in `{{mode_info}}` and `{{context_info}}`, making it impossible for me to properly customize the statusline, without removing information.

I would like to propose adding a few more value expansions, so that (power) users are able to customize the statusline without compromising on the shown information.
As you can see below, there are a few values that can be argued over (selection_index and num_selections), as you can probably already compute them with existing values. However I think it is more user friendly to add these values.
By adding these extra values, we effectively allow users to recreate the existing status_line without `{{mode_info}}` and `{{context_info}}`.

<details>
<summary>Value Expansions</summary>

_The names chosen below are open for discussion._

<dl>
  <dt>buftype</dt>
  <dd>

The type of the buffer (file, scratch, debug or fifo).

  </dd>

  <dt>selection_index</dt>
  <dd>

The index of the main selection. You can probably already get this from `%val{selection}` and `%val{selections}`, but it is quite cumbersome to do so.

</dd>

  <dt>selection_count</dt>
  <dd>

The number of selections. Again, this can be computed from `%val{selections}` but it is cumbersome to do.

  </dd>

  <dt>recording_reg</dt>
  <dd>

The register which is currently being used to record a macro.

  </dd>

  <dt>current_reg</dt>
  <dd>

_(normal mode only)_ Currently selected register (`NormalMode.m_params.reg`).

</dd>

  <dt>current_count</dt>
  <dd>

_(normal mode only)_ Current param number (`NormalMode.m_params.count`).

  </dd>

  <dt>mode</dt>
  <dd>

The mode we are currently in.

  </dd>

  <dt>new</dt>
  <dd>

Whether the buffer is a new file.

  </dd>

  <dt>read_only</dt>
  <dd>

Whether the file is read only.

  </dd>

  <dt>hooks_enabled</dt>
  <dd>

Whether hooks are enabled

  </dd>
</dl>
</details>

In #1171, some users complained that the current mode was not very visible in the current status line.
While this issue has been closed, I still feel that the current `{{mode_info}}` is quite cluttered.
Adding a `promptfmt` which works similarly to the `modelinefmt`, but is added to the left (before the statusline), would allow users to eg. show the current mode on the left.

---

As this is my first time working with Kakoune, I already implemented these changes in [my own fork](https://github.com/0phoff/kakoune/tree/dev-statusline), to make sure I could actually add these things and submit a PR if this issue gets accepted.

I you want to have this functionality, let me know and I will open a PR from my branch.
I have a few technical questions about the code and might need some help with Testing / Documentation, but as far as I am aware, my implementation works correctly!

**Comment by raiguard** [comment]
I would absolutely love these capabilities, and I would support a PR. Especially `promptfmt`, that would make it infinitely easier to see what mode I am in.

**Comment by 0phoff** [comment]
> I would absolutely love these capabilities, and I would support a PR. Especially `promptfmt`, that would make it infinitely easier to see what mode I am in.

I will rebase my changes on the latest master branch and create a PR.
I would need some help with unit testing, as I am not familiar with how kakoune handles this! :innocent:

**Comment by krobelus** [comment]
I don't care much about this feature but it sounds reasonable to expose this information.

> However, Kakoune encodes a lot of very useful information in {{mode_info}} and {{context_info}}, making it impossible for me to properly customize the statusline, without removing information.

Random thought: what if this were available as `%val{mode_info}`, then users could parse it, or it could be a list, then users just need to index it to get what they want. Probably your approach is better.

> As you can see below, there are a few values that can be argued over (selection_index and num_selections), as you can > probably already compute them with existing values. However I think it is more user friendly to add these values.

`selection_count` can probably go; it's fairly easy to compute with `set -- $kak_selections_desc; echo $#`.
`%val{read_only}` should probably be the existing `%opt{readonly}` (I didn't check)

> Config

That's a lot of shell calls. Not sure if it matters but [here](https://github.com/Delapouite/dot-in-the-sky/blob/924c718c07a433b1e7e496bbea3daed64ebed314/.config/kak/kakrc#L12-L29) is an example that only needs one shell call per redraw (?).

**Comment by krobelus** [comment]
> I would need some help with unit testing, as I am not familiar with how kakoune handles this!

Check `test/README.asciidoc` and observe that `test/**/script` contain `ui_out` assertions that test modeline rendering. I think you can do the same.

**Comment by raiguard** [comment]
Yes, `%val{read_only}` is always 1:1 with `%opt{readonly}` AFAIK, so that can be removed.

[Here](https://github.com/raiguard/dotfiles/blob/master/.config/kak/tool/status-line.kak) is my implementation using these features. I originally did something very similar to 0phoff, but quickly had performance issues. Combined with PR #4509, I use several string options to update the content, and only do so when needed using a variety of hooks.

(also thanks 0phoff for showing your "dim braces" style, I really liked it and ~~stole~~assimilated it!)

![image](https://user-images.githubusercontent.com/3515394/150656639-1054df9b-62c0-4461-9220-43489973e6b7.png)

**Comment by ismay** [comment]
This looks great, exactly what I was looking for. One important thing about customizing the modeline in my opinion is that, as soon as there's information you care about that's only in the mode_info or context_info, you're stuck with using them for _everything_ that they expose. After all, as soon as you're using them there's no use in duplicating the information that they show.

So effectively they monopolize a bunch of information that I'd personally want to show differently, but can't. The proposal here sounds like a good way to solve it to me.

**Comment by mawww** [comment]
303d87978556b3548ca4bab98f5daba7f804e4b8 made a step in that direction as we now return a struct from the mode_info() method on modes, I think all of the information provided by modes is now available except for the mode name, which could be provided in that struct.

The current behaviour of `{{mode_info}}` cannot be fully replaced by this though because we would need to shell out to decide whether to add the selection info part (selection count + main index).
